main/src/ca/uwaterloo/flix/language/phase/unification/FastSetUnification.scala

```diff
@@ -0,0 +1,338 @@
+      * [[Compl]], [[Union]], and [[Inter]] are connectives.
+      */
+    final def size: Int = {
+      var workList = List(this)
+      var counter = 0
+
+      /** Updates `counter` and `workList` given intersection or union subformulas. */
+      @inline
+      def countSetFormulas(
+                            elemPos: Option[ElemSet], cstsPos: Set[Cst], varsPos: Set[Var],
+                            elemNeg: Option[ElemSet], cstsNeg: Set[Cst], varsNeg: Set[Var],
+                            other: List[SetFormula]
+                          ): Unit = {
+        val negElemSize = elemNeg.size
+        val negCstsSize = cstsNeg.size
+        val negVarsSize = varsNeg.size
+        val subformulas = elemPos.size + cstsPos.size + varsPos.size +
+          negElemSize + negCstsSize + negVarsSize + other.size
+        val connectives = negElemSize + negCstsSize + negVarsSize
+        counter += (subformulas - 1) + connectives
+        workList = other ++ workList
+      }
+
+      while (workList.nonEmpty) {
+        val t :: next = workList
+        workList = next
+        t match {
+          case Univ => ()
+          case Empty => ()
+          case Cst(_) => ()
+          case Var(_) => ()
+          case ElemSet(_) => ()
+          case Compl(t) =>
+            counter += 1
+            workList = t :: workList
+          case Inter(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other) =>
+            countSetFormulas(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other)
+          case Union(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other) =>
+            countSetFormulas(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other)
+        }
+      }
+      counter
+    }
+
+    /** Returns a human-readable string of `this`. */
+    override final def toString: String = this match {
+      case Univ => "univ"
+      case Empty => "empty"
+      case Cst(c) => s"c$c"
+      case ElemSet(s) if s.sizeIs == 1 => s"e${s.head}"
+      case ElemSet(s) => s"e${s.mkString("+")}"
+      case Var(x) => s"x$x"
+      case Compl(f) => f match {
+        case Univ | Empty | Cst(_) | ElemSet(_) | Var(_) | Compl(_) => s"!$f"
+        case Inter(_, _, _, _, _, _, _) | Union(_, _, _, _, _, _, _) => s"!($f)"
+      }
+      case Inter(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other) =>
+        val subformulas = subformulasOf(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other)
+        s"(${subformulas.mkString(" ∩ ")})"
+      case Union(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other) =>
+        val subformulas = subformulasOf(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other)
+        s"(${subformulas.mkString(" ∪ ")})"
+    }
+
+  }
+
+  object SetFormula {
+
+    /** Skip invariant checks if `false`. */
+    private val CHECK_INVARIANTS: Boolean = true
+
+    /** A trait used to refer to [[Univ]] and [[Empty]] collectively. */
+    sealed trait UnivOrEmpty extends SetFormula
```

Comment:
How is this later used?

Reply:
In propagation, there is a partial substitution only to Univ or Empty so its just for added type safety / pattern match exhaustiveness